### PR TITLE
refactor(cardinal): register query with generics

### PR DESCRIPTION
**Pull request by smsunarto**
Closes: #XXX

## What is the purpose of the change

Making register query more consistent with register component

## Brief Changelog

- Refactor RegisterQuery to be similar to RegisterComponent

## Testing and Verifying

- All relevant tests have been adjusted accordingly

### CODE REVIEW

**Comment by smsunarto** [comment]
Current dependencies on/for this PR:
* main
  * **PR #412** <a href="https://app.graphite.dev/github/pr/Argus-Labs/world-engine/412" target="_blank">Graphite</a> 
    * **PR #413** <a href="https://app.graphite.dev/github/pr/Argus-Labs/world-engine/413" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/Argus-Labs/world-engine/413?utm_source=stack-comment).

File: cardinal/ecs/query_test.go [hunk 1]
```diff
@@ -11,41 +11,14 @@ import (
...
-		}()
-	}()
-
-	ecs.NewQueryType[string, string]("foo", func(wCtx ecs.WorldContext, req string) (string, error) {
-		return "blah", nil
+	str := "blah"
+	err := ecs.RegisterQuery[string, string](testutils.NewTestWorld(t).Instance(), "foo", func(wCtx ecs.WorldContext, req *string) (*string, error) {
+		return &str, nil
 	})
+	assert.Assert(t, err != nil)
```
Comment (technicallyty) [inline cardinal/ecs/query_test.go:21]: we should ensure this error is hitting the right case.

i'd suggest using `assert.ErrorContains` here and make sure it has the `The Request and Reply must be both structs` string.

File: cardinal/ecs/query.go [hunk 2]
```diff
@@ -217,3 +197,32 @@ func (r *QueryType[Request, Reply]) EncodeAsABI(input any) ([]byte, error) {
...
+	}
+	repType := reflect.TypeOf(rep)
+	repKind := reqType.Kind()
+	repValid := false
+	if (repKind == reflect.Pointer && repType.Elem().Kind() == reflect.Struct) || repKind == reflect.Struct {
+		repValid = true
+	}
+
+	if !repValid || !reqValid {
+		return errors.New(fmt.Sprintf("Invalid Query: %s: The Request and Reply must be both structs", name))
```
Comment (technicallyty) [inline cardinal/ecs/query.go:225]: go errors should not start with a capital:
https://google.github.io/styleguide/go/decisions.html#:~:text=Error%20strings%20should%20not%20be,being%20printed%20to%20the%20user.

File: cardinal/ecs/query_test.go [hunk 3]
```diff
@@ -101,31 +76,32 @@ func TestPanicsOnNoNameOrHandler(t *testing.T) {
 	type foo struct{}
 	testCases := []struct {
 		name        string
-		createQuery func()
-		shouldPanic bool
+		createQuery func() error
+		shouldErr   bool
 	}{
 		{
 			name: "panic on no name",
```
Comment (technicallyty) [inline cardinal/ecs/query_test.go:83]: should update these test names as well

Comment (technicallyty) [reply cardinal/ecs/query_test.go:93]: looks like this was marked as resolved but wasn't fixed

File: cardinal/ecs/query_test.go [hunk 4]
```diff
@@ -97,35 +82,36 @@ func TestQueryEVM(t *testing.T) {
...
-			createQuery: func() {
-				ecs.NewQueryType[foo, foo]("", nil)
+			createQuery: func() error {
+				return ecs.RegisterQuery[foo, foo](testutils.NewTestWorld(t).Instance(), "", nil)
 			},
-			shouldPanic: true,
+			shouldErr: true,
 		},
 		{
 			name: "panic on no handler",
```
Comment (technicallyty) [inline cardinal/ecs/query_test.go:100]: update name here

**Comment by smsunarto** [comment]
### Merge activity

* **Nov 13, 3:48 PM**: [Graphite](https://app.graphite.dev/github/pr/Argus-Labs/world-engine/413) rebased this pull request after merging its parent, because this pull request is set to merge when ready.

**Review by jerargus — APPROVED** [review]

**Review by technicallyty — APPROVED** [review]

**Review by technicallyty — APPROVED** [review]

**Comment by smsunarto** [comment]
gdi, stop merging in front of me

**Comment by smsunarto** [comment]
fuck e2e tests, all my homies hate the e2e

**Pull request merged**